src/appleseed-max-impl/appleseedosltexture/oslshadermetadata.h

```diff
@@ -0,0 +1,196 @@
+#include "foundation/core/concepts/noncopyable.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// 3ds Max headers.
+#include <maxapi.h>
+#include <maxtypes.h>
+
+// Standard headers.
+#include <vector>
+#include <map>
+#include <string>
+
+// Forward declarations.
+namespace renderer { class ShaderQuery; }
+class OSLParamInfo;
+
+typedef std::map<int, const std::wstring> IdNameMap;
+typedef std::vector<std::pair<int, std::wstring>> IdNameVector;
+
+class MaxParam
+{
+  public:
+    enum ParamType
+    {
+        Float,
+        Float2f,
+        VectorParam,
+        NormalParam,
+        PointParam,
+        Matrix,
+        IntNumber,
+        IntCheckbox,
+        IntMapper,
+        StringPopup,
+        Color,
+        Closure,
+        String,
+        Unsupported
+    };
+
+    bool is_vector() const
+    {
+        return  param_type == VectorParam ||
+                param_type == NormalParam ||
+                param_type == PointParam;
+    }
+
+    ParamType       param_type;
+    bool            connectable;
+    std::string     max_label_str;
+    int             max_param_id;
+    int             max_ctrl_id;
+    std::string     osl_param_name;
+    std::string     page_name;
+};
+
+
+//
+// The OSLMetadataExtractor class extracts OSL metadata entries from appleseed dictionaries.
+//
+
+class OSLMetadataExtractor
+    : public foundation::NonCopyable
+{
+public:
+    explicit OSLMetadataExtractor(const foundation::Dictionary& metadata);
+
+    bool exists(const char* key) const;
+
+    bool getValue(const char* key, std::string& value);
+
+    bool getValue(const char* key, bool& value);
+
+    template <typename T>
+    bool getValue(const char* key, T& value) const
+    {
+        if (exists(key))
+        {
+            const foundation::Dictionary& dict = m_metadata.dictionary(key);
+            value = dict.get<T>("value");
+            return true;
+        }
+
+        return false;
+    }
+
+private:
+    const foundation::Dictionary& m_metadata;
+};
+
+//
+// The OSLParamInfo class holds information about a parameter of an OSL shader.
+//
+
+class OSLParamInfo
+{
+  public:
+    explicit OSLParamInfo(const foundation::Dictionary& paramInfo);
+
+    // Query info.
+    std::string paramName;
+    std::string paramType;
+    bool isOutput;
+    bool isClosure;
+    bool isStruct;
+    std::string structName;
+    bool isArray;
+    int arrayLen;
+    bool lockGeom;
+
+    // Defaults.
+    bool validDefault;
+    bool hasDefault;
+    std::vector<double> defaultValue;
+    std::string defaultStringValue;
+
+    // Standard metadata info.
+    std::string units;
+    std::string page;
+    std::string label;
+    std::string widget;
+    std::string options;
+    std::string help;
+    bool hasMin;
+    double minValue;
+    bool hasMax;
+    double maxValue;
+    bool hasSoftMin;
+    double softMinValue;
+    bool hasSoftMax;
+    double softMaxValue;
+    bool divider;
+
+    bool mayaAttributeConnectable;
```

Comment:
I'd use a more generic name, like connectable.
We use right now the maya_attribute_connectable metadata, but there is a new standard metadata entry for this in newer OSL versions and we'll switch to that probably soon.